src/routes/publish.ts

```diff
@@ -8,196 +8,602 @@ import { StatsWalesApi } from '../services/stats-wales-api';
+        );
         return;
     }
+    const statsWalesApi = new StatsWalesApi(lang, req.jwt);
     if (confirmData === 'true') {
+        logger.info('User confirmed file upload was correct');
+        await confirmFileUpload(currentDataset, currentRevision, currentFileImport, statsWalesApi, req, res);
+    } else {
+        logger.info('User rejected the file in preview');
+        await rejectFileReturnToUpload(currentDataset, currentRevision, currentFileImport, statsWalesApi, req, res);
+    }
+});
+
+function updateCurrentImport(currentImport: ImportDTO, dimensionCreationRequest: DimensionCreationDTO[]) {
+    if (currentImport.sources) {
+        currentImport.sources.forEach((source) => {
+            source.type =
+                dimensionCreationRequest.find((dim) => dim.sourceId === source.id)?.sourceType || SourceType.UNKNOWN;
+        });
+    }
+    return currentImport;
+}
+
+publish.get('/sources', upload.none(), (req: AuthedRequest, res: Response) => {
+    const lang = req.i18n.language;
+    let currentFileImport = checkCurrentFileImport(req, res);
+    const dimensionCreationRequest = req.session.dimensionCreationRequest;
+    if (!currentFileImport) {
+        return;
+    }
+    if (currentFileImport.sources.length === 0) {
+        logger.error('No current import found in the session with sources... user may have navigated here by mistake');
+        req.session.errors = generateViewErrors(undefined, 500, [
+            generateError('session', 'errors.session.no_sources_on_import', {})
+        ]);
+        req.session.save();
+        res.redirect(`/${lang}/${req.i18n.t('routes.publish.start', { lng: lang })}`);
+        return;
+    }
+    const errs = req.session.errors;
+    req.session.errors = undefined;
+    req.session.save();
+    if (errs) {
+        res.status(500);
+    } else {
         res.status(200);
-        res.json({ confirm: true, dataset_id: dataset.id, message: 'Datatable has been confirmed as correct' });
+    }
+
+    if (dimensionCreationRequest) {
+        currentFileImport = updateCurrentImport(currentFileImport, dimensionCreationRequest);
     } else {
+        currentFileImport.sources.forEach((source) => {
+            source.type = SourceType.UNKNOWN;
+        });
+    }
+    res.render('publish/sources', {
+        errors: errs,
+        currentImport: currentFileImport,
+        sourceTypes: Object.keys(SourceType)
+    });
+});
+
+publish.post('/sources', upload.none(), async (req: AuthedRequest, res: Response) => {
+    const lang = req.i18n.language;
+    const currentDataset = checkCurrentDataset(req, res);
+    if (!currentDataset) {
+        return;
+    }
+
+    const currentRevision = checkCurrentRevision(req, res);
+    if (!currentRevision) {
+        return;
+    }
+
+    const currentFileImport = checkCurrentFileImport(req, res);
+    if (!currentFileImport) {
+        return;
+    }
```

Comment:
So I think you have to have the checks because its TypeScript... If the object could be undefined (which it can be when getting it from the session) you have to confirm the object you want to use actually exists otherwise the typescript compiler will throw errors at you.

Reply:
That can be resolved with typing - before the middleware is run your session type props are optional. Then the middleware throw exceptions if they're missing. Once the middleware has run, you use a type that removes optionality from those props.